Reading error at high transaction loads

When running at higher transaction loads (multiple per second) a transaction would sometimes fail with a read error, e.g.
```
[INFO] [02/18/2015 15:20:16.086] [mediator-akka.actor.default-dispatcher-12] [akka://mediator/user/xds-mediator/$5c] Received request: POST /xdsrepository
[Fatal Error] :54:19: The element type "rim:ExtrinsicObject" must be terminated by the matching end-tag "</rim:ExtrinsicObject>".
[ERROR] [02/18/2015 15:20:16.086] [mediator-akka.actor.default-dispatcher-15] [akka://mediator/user/xds-mediator/$5c] Exception while processing request
org.xml.sax.SAXParseException; lineNumber: 54; columnNumber: 19; The element type "rim:ExtrinsicObject" must be terminated by the matching end-tag "</rim:ExtrinsicObject>".
	at com.sun.org.apache.xerces.internal.parsers.DOMParser.parse(DOMParser.java:257)
	at com.sun.org.apache.xerces.internal.jaxp.DocumentBuilderImpl.parse(DocumentBuilderImpl.java:348)
	at javax.xml.parsers.DocumentBuilder.parse(DocumentBuilder.java:121)
	at org.openhim.mediator.orchestration.RepositoryActor.getSOAPActionFromHeader(RepositoryActor.java:118)
	at org.openhim.mediator.orchestration.RepositoryActor.determineSOAPAction(RepositoryActor.java:95)
	at org.openhim.mediator.orchestration.RepositoryActor.triggerRepositoryAction(RepositoryActor.java:180)
	at org.openhim.mediator.orchestration.RepositoryActor.readMessage(RepositoryActor.java:76)
	at org.openhim.mediator.orchestration.RepositoryActor.onReceive(RepositoryActor.java:226)
	at akka.actor.UntypedActor$$anonfun$receive$1.applyOrElse(UntypedActor.scala:167)
	at akka.actor.Actor$class.aroundReceive(Actor.scala:465)
	at akka.actor.UntypedActor.aroundReceive(UntypedActor.scala:97)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:516)
	at akka.actor.ActorCell.invoke(ActorCell.scala:487)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:254)
	at akka.dispatch.Mailbox.run(Mailbox.scala:221)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:231)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)

[ERROR] [02/18/2015 15:20:48.140] [mediator-akka.actor.default-dispatcher-26] [akka://mediator/user/xds-mediator] Request containment exception
java.io.IOException: java.util.concurrent.TimeoutException
	at org.glassfish.grizzly.nio.transport.TCPNIOTransportFilter.handleRead(TCPNIOTransportFilter.java:90)
	at org.glassfish.grizzly.filterchain.TransportFilter.handleRead(TransportFilter.java:173)
	at org.glassfish.grizzly.filterchain.ExecutorResolver$9.execute(ExecutorResolver.java:119)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeFilter(DefaultFilterChain.java:283)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeChainPart(DefaultFilterChain.java:200)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.read(DefaultFilterChain.java:351)
	at org.glassfish.grizzly.filterchain.FilterChainContext.read(FilterChainContext.java:736)
	at org.glassfish.grizzly.http.io.InputBuffer.blockingRead(InputBuffer.java:1119)
	at org.glassfish.grizzly.http.server.io.ServerInputBuffer.blockingRead(ServerInputBuffer.java:95)
	at org.glassfish.grizzly.http.io.InputBuffer.fillChars(InputBuffer.java:1223)
	at org.glassfish.grizzly.http.io.InputBuffer.read(InputBuffer.java:490)
	at org.glassfish.grizzly.http.io.InputBuffer.read(InputBuffer.java:553)
	at org.glassfish.grizzly.http.server.NIOReaderImpl.read(NIOReaderImpl.java:82)
	at org.apache.commons.io.IOUtils.copyLarge(IOUtils.java:2001)
	at org.apache.commons.io.IOUtils.copyLarge(IOUtils.java:1980)
	at org.apache.commons.io.IOUtils.copy(IOUtils.java:1957)
	at org.apache.commons.io.IOUtils.toString(IOUtils.java:819)
	at org.openhim.mediator.engine.MediatorRootActor.buildMediatorHTTPRequestFromGrizzlyRequest(MediatorRootActor.java:115)
	at org.openhim.mediator.engine.MediatorRootActor.access$000(MediatorRootActor.java:45)
	at org.openhim.mediator.engine.MediatorRootActor$1.call(MediatorRootActor.java:83)
	at akka.dispatch.Futures$$anonfun$future$1.apply(Future.scala:94)
	at scala.concurrent.impl.Future$PromiseCompletingRunnable.liftedTree1$1(Future.scala:24)
	at scala.concurrent.impl.Future$PromiseCompletingRunnable.run(Future.scala:24)
	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:41)
	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:401)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
Caused by: java.util.concurrent.TimeoutException
	at org.glassfish.grizzly.nio.tmpselectors.TemporarySelectorReader.read(TemporarySelectorReader.java:126)
	at org.glassfish.grizzly.nio.tmpselectors.TemporarySelectorReader.read(TemporarySelectorReader.java:75)
	at org.glassfish.grizzly.AbstractReader.read(AbstractReader.java:72)
	at org.glassfish.grizzly.nio.transport.TCPNIOTransportFilter.handleRead(TCPNIOTransportFilter.java:77)
	... 28 more
```

--- a/src/main/java/org/openhim/mediator/engine/MediatorRootActor.java
+++ b/src/main/java/org/openhim/mediator/engine/MediatorRootActor.java
@@ -6,29 +6,26 @@
 
 package org.openhim.mediator.engine;
 
-import static akka.dispatch.Futures.future;
-
 import akka.actor.*;
 import akka.dispatch.OnComplete;
 import akka.event.Logging;
 import akka.event.LoggingAdapter;
 import org.apache.commons.io.IOUtils;
-import org.glassfish.grizzly.Buffer;
 import org.glassfish.grizzly.ReadHandler;
-import org.glassfish.grizzly.http.io.NIOInputStream;
 import org.glassfish.grizzly.http.io.NIOReader;
 import org.glassfish.grizzly.http.server.Response;
 import org.openhim.mediator.engine.connectors.CoreAPIConnector;
 import org.openhim.mediator.engine.connectors.HTTPConnector;
 import org.openhim.mediator.engine.connectors.MLLPConnector;
 import org.openhim.mediator.engine.connectors.UDPFireForgetConnector;
-import org.openhim.mediator.engine.messages.*;
+import org.openhim.mediator.engine.messages.GrizzlyHTTPRequest;
+import org.openhim.mediator.engine.messages.MediatorHTTPRequest;
+import org.openhim.mediator.engine.messages.MediatorHTTPResponse;
+import org.openhim.mediator.engine.messages.RegisterMediatorWithCore;
 import scala.concurrent.ExecutionContext;
 import scala.concurrent.Future;
 import scala.concurrent.duration.Duration;
 import scala.concurrent.duration.FiniteDuration;
-
-import java.io.ByteArrayOutputStream;
 import java.io.IOException;
 import java.io.StringWriter;
 import java.util.HashMap;
@@ -36,6 +33,8 @@ import java.util.Map;
 import java.util.TreeMap;
 import java.util.concurrent.Callable;
 import java.util.concurrent.TimeUnit;
+
+import static akka.dispatch.Futures.future;
 
 /**
  * The root actor for the mediator.
@@ -129,7 +128,6 @@ public class MediatorRootActor extends UntypedActor {
             params.put(param, request.getRequest().getParameter(param));
         }
 
-        //handler for reading data
         in.notifyAvailable(new ReadHandler() {
             final StringWriter bodyBuffer = new StringWriter();
             char[] readBuffer = new char[1024];
@@ -152,10 +150,10 @@ public class MediatorRootActor extends UntypedActor {
             @Override
             public void onError(Throwable throwable) {
                 try {
-                    log.error(throwable, "Request containment exception");
+                    log.error(throwable, "Error during reading of request body");
                     handleResponse(request.getResponseHandle(), 500, "text/plain", throwable.getMessage());
                 } catch (IOException ex) {
-                    log.error(ex, "Exception during onError");
+                    log.error(ex, "Error during reading of request body");
                 } finally {
                     request.getResponseHandle().resume();
                 }
@@ -163,27 +161,32 @@ public class MediatorRootActor extends UntypedActor {
 
             @Override
             public void onAllDataRead() throws Exception {
-                read();
-
-                MediatorHTTPRequest mediatorHTTPRequest = new MediatorHTTPRequest(
-                        requestHandler,
-                        requestHandler,
-                        null,
-                        request.getRequest().getMethod().toString(),
-                        request.getRequest().getScheme(),
-                        request.getRequest().getLocalAddr(),
-                        request.getRequest().getLocalPort(),
-                        request.getRequest().getRequestURI(),
-                        bodyBuffer.toString(),
-                        headers,
-                        params
-                );
-
-                handlerInbox.send(requestHandler, mediatorHTTPRequest);
-
-                IOUtils.closeQuietly(in);
+                try {
+                    read();
+                    MediatorHTTPRequest mediatorHTTPRequest = buildMediatorHTTPRequest(requestHandler, request, bodyBuffer.toString(), headers, params);
+                    handlerInbox.send(requestHandler, mediatorHTTPRequest);
+                } finally {
+                    IOUtils.closeQuietly(in);
+                }
             }
         });
+    }
+
+    private MediatorHTTPRequest buildMediatorHTTPRequest(ActorRef requestHandler, GrizzlyHTTPRequest request,
+                                                         String body, Map<String, String> headers, Map<String, String> params) {
+        return new MediatorHTTPRequest(
+                requestHandler,
+                requestHandler,
+                null,
+                request.getRequest().getMethod().toString(),
+                request.getRequest().getScheme(),
+                request.getRequest().getLocalAddr(),
+                request.getRequest().getLocalPort(),
+                request.getRequest().getRequestURI(),
+                body,
+                headers,
+                params
+        );
     }
 
     private void handleResponse(Response grizzlyResponseHandle, MediatorHTTPResponse response) throws IOException {

--- a/src/main/java/org/openhim/mediator/engine/MediatorRootActor.java
+++ b/src/main/java/org/openhim/mediator/engine/MediatorRootActor.java
@@ -13,6 +13,10 @@ import akka.dispatch.OnComplete;
 import akka.event.Logging;
 import akka.event.LoggingAdapter;
 import org.apache.commons.io.IOUtils;
+import org.glassfish.grizzly.Buffer;
+import org.glassfish.grizzly.ReadHandler;
+import org.glassfish.grizzly.http.io.NIOInputStream;
+import org.glassfish.grizzly.http.io.NIOReader;
 import org.glassfish.grizzly.http.server.Response;
 import org.openhim.mediator.engine.connectors.CoreAPIConnector;
 import org.openhim.mediator.engine.connectors.HTTPConnector;
@@ -24,7 +28,9 @@ import scala.concurrent.Future;
 import scala.concurrent.duration.Duration;
 import scala.concurrent.duration.FiniteDuration;
 
+import java.io.ByteArrayOutputStream;
 import java.io.IOException;
+import java.io.StringWriter;
 import java.util.HashMap;
 import java.util.Map;
 import java.util.TreeMap;
@@ -80,9 +86,8 @@ public class MediatorRootActor extends UntypedActor {
 
         Future<Object> f = future(new Callable<Object>() {
             public Object call() throws IOException {
-                MediatorHTTPRequest mediatorHTTPRequest = buildMediatorHTTPRequestFromGrizzlyRequest(requestHandler, request);
                 Inbox inbox = Inbox.create(getContext().system());
-                inbox.send(requestHandler, mediatorHTTPRequest);
+                processGrizzlyRequest(inbox, requestHandler, request);
                 return inbox.receive(getRootTimeout());
             }
         }, ec);
@@ -111,32 +116,74 @@ public class MediatorRootActor extends UntypedActor {
     }
 
 
-    private MediatorHTTPRequest buildMediatorHTTPRequestFromGrizzlyRequest(ActorRef requestHandler, GrizzlyHTTPRequest request) throws IOException {
-        String body = IOUtils.toString(request.getRequest().getNIOReader());
+    private void processGrizzlyRequest(final Inbox handlerInbox, final ActorRef requestHandler, final GrizzlyHTTPRequest request) throws IOException {
+        final NIOReader in = request.getRequest().getNIOReader();
 
-        Map<String, String> headers = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
+        final Map<String, String> headers = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
         for (String hdr : request.getRequest().getHeaderNames()) {
             headers.put(hdr, request.getRequest().getHeader(hdr));
         }
 
-        Map<String, String> params = new HashMap<>();
+        final Map<String, String> params = new HashMap<>();
         for (String param : request.getRequest().getParameterNames()) {
             params.put(param, request.getRequest().getParameter(param));
         }
 
-        return new MediatorHTTPRequest(
-                requestHandler,
-                requestHandler,
-                null,
-                request.getRequest().getMethod().toString(),
-                request.getRequest().getScheme(),
-                request.getRequest().getLocalAddr(),
-                request.getRequest().getLocalPort(),
-                request.getRequest().getRequestURI(),
-                body,
-                headers,
-                params
-        );
+        //handler for reading data
+        in.notifyAvailable(new ReadHandler() {
+            final StringWriter bodyBuffer = new StringWriter();
+            char[] readBuffer = new char[1024];
+
+            private void read() throws IOException {
+                while (in.isReady()) {
+                    int len = in.read(readBuffer);
+                    if (len > 0) {
+                        bodyBuffer.write(readBuffer, 0, len);
+                    }
+                }
+            }
+
+            @Override
+            public void onDataAvailable() throws Exception {
+                read();
+                in.notifyAvailable(this);
+            }
+
+            @Override
+            public void onError(Throwable throwable) {
+                try {
+                    log.error(throwable, "Request containment exception");
+                    handleResponse(request.getResponseHandle(), 500, "text/plain", throwable.getMessage());
+                } catch (IOException ex) {
+                    log.error(ex, "Exception during onError");
+                } finally {
+                    request.getResponseHandle().resume();
+                }
+            }
+
+            @Override
+            public void onAllDataRead() throws Exception {
+                read();
+
+                MediatorHTTPRequest mediatorHTTPRequest = new MediatorHTTPRequest(
+                        requestHandler,
+                        requestHandler,
+                        null,
+                        request.getRequest().getMethod().toString(),
+                        request.getRequest().getScheme(),
+                        request.getRequest().getLocalAddr(),
+                        request.getRequest().getLocalPort(),
+                        request.getRequest().getRequestURI(),
+                        bodyBuffer.toString(),
+                        headers,
+                        params
+                );
+
+                handlerInbox.send(requestHandler, mediatorHTTPRequest);
+
+                IOUtils.closeQuietly(in);
+            }
+        });
     }
 
     private void handleResponse(Response grizzlyResponseHandle, MediatorHTTPResponse response) throws IOException {

--- a/src/main/java/org/openhim/mediator/engine/MediatorServer.java
+++ b/src/main/java/org/openhim/mediator/engine/MediatorServer.java
@@ -47,8 +47,6 @@ public class MediatorServer {
 
     private void configureHttpServer() {
         NetworkListener listener = new NetworkListener(config.getName(), config.getServerHost(), config.getServerPort());
-        //Disabled grizzly's thread pooling, we will handle our own threading (Akka)
-        listener.getTransport().setWorkerThreadPoolConfig(null);
         httpServer.addListener(listener);
 
         httpServer.getServerConfiguration().addHttpHandler(new HttpHandler() {

--- a/src/test/java/org/openhim/mediator/engine/MediatorServerTest.java
+++ b/src/test/java/org/openhim/mediator/engine/MediatorServerTest.java
@@ -124,12 +124,16 @@ public class MediatorServerTest {
 
     private static class POSTBigMediatorActor extends UntypedActor {
         public static String TEST_MESSAGE;
+        public static int ROGUE_B_POS = 1024*1024*7 + 1234;
 
         static {
             StringBuilder msg = new StringBuilder();
             for (int i=0; i<10*1024*1024; i++) {
                 msg.append('a');
             }
+            //set a rogue 'b' for validation
+            msg.setCharAt(ROGUE_B_POS, 'b');
+
             TEST_MESSAGE = msg.toString();
         }
 
@@ -138,6 +142,7 @@ public class MediatorServerTest {
             if (msg instanceof MediatorHTTPRequest) {
                 assertEquals(TEST_MESSAGE.length(), ((MediatorHTTPRequest) msg).getBody().length());
                 assertEquals(TEST_MESSAGE, ((MediatorHTTPRequest) msg).getBody());
+                assertEquals('b', ((MediatorHTTPRequest) msg).getBody().charAt(ROGUE_B_POS));
 
                 FinishRequest fr = new FinishRequest(null, "text/plain", 201);
                 ((MediatorHTTPRequest) msg).getRequestHandler().tell(fr, getSelf());

